### PR TITLE
NNCF task

### PR DESCRIPTION
**Pull request by alexsu52**
Change list:
- Split the OTEClassificationTask into OTEClassificationInferenceTask and OTEClassificationTrainingTask
- Added support for NNCF scenarios in the ote_sample
- Added OTEClassificationNNCFTask
- Added compression_configs for the ote_custom classification templates
- Refactored the config machinery 

The following results were obtained on the CIFAR100:
| model name | FP32 Acc@1 | NNCF INT8 Acc@1 | OpenVINO INT8 Acc@1 | SpeedUp | NNCF training (drop <= 1) |
|--|:--:|:--:|:--:|:--:|:--:|
| effecienet_b0 | 85.82 | 84.98 | 85.03 | 1.34 | initialization |
| mobilenet_v3_large | 85.23 | 84.34 | 84.33 | 1.52 | initialization |
| mobilenet_v3_large_075 | 84.9 | 84.03 | 84.02 | 1.42 | training (1 epochs) |
| mobilenet_v3_small | | | | | |

The following results were obtained on the VOC for the multilabel scenario:

| model name | FP32 Acc@1 | NNCF INT8 Acc@1 | OpenVINO INT8 Acc@1 | SpeedUp | NNCF training (drop <= 1) |
|--|:--:|:--:|:--:|:--:|:--:|
| effecienet_b0 | 88.43 (97.7) | 88.45 (97.73) | - (97.74) | 1.34 | initialization |
| mobilenet_v3_large_075 | 90.07 (97.88) | 89.08 (97.78) | - (97.78) | 1.42 | training (4 epochs) |

### CODE REVIEW

File: configs/ote_custom_classification/mobilenet_v3_large_075/template.yaml [hunk 1]
```diff
@@ -15,6 +15,7 @@ framework: OTEClassification v1.2.3
 entrypoints:
   base: torchreid.integration.sc.task.OTEClassificationTask
```
Comment (sovrasov) [inline configs/ote_custom_classification/mobilenet_v3_large_075/template.yaml]: Entry point should be changed to `OTEClassificationTrainTask`

Comment (alexsu52) [reply configs/ote_custom_classification/mobilenet_v3_large_075/template.yaml]: Done

File: configs/ote_custom_classification/efficientnet_b0/compression_config.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{
+    "base": {
+        "nncf_config": {
+            "target_metric_name": "mAP",
```
Comment (kprokofi) [inline configs/ote_custom_classification/efficientnet_b0/compression_config.json]: why mAP?

File: tools/ote_sample.py [hunk 3]
```diff
@@ -73,33 +98,64 @@ def main(args):
...
     params = create(model_template.hyper_parameters.data)
     logger.info('Setup environment')
-    environment = TaskEnvironment(model=None, hyper_parameters=params, label_schema=labels_schema, model_template=model_template)
+    environment = TaskEnvironment(model=None,
+                                  hyper_parameters=params,
+                                  label_schema=labels_schema,
+                                  model_template=model_template)
 
-    logger.info('Create base Task')
+    training_from_scratch = args.weights is None
```
Comment (kprokofi) [inline tools/ote_sample.py]: maybe not from scratch, but 'validation_only' or something like this. Because loaded weights can be used further in training. Here, as I see, we only validate the model if weights are presented. FMO, It is better to treat it as a validation-only process and redo this a bit.

File: torchreid/__init__.py [hunk 4]
```diff
@@ -1,5 +1,16 @@
 from __future__ import absolute_import, print_function
 
+try:
```
Comment (kprokofi) [inline torchreid/__init__.py:3]: I guess this check is useless since nothing will work without torch as well as without all other requirements from the requirements.txt

File: torchreid/apis/training.py [hunk 5]
```diff
@@ -100,14 +108,25 @@ def run_training(cfg, datamanager, model, optimizer, scheduler, extra_device_ids
                           initial_lr=init_lr)
 
     accuracy = None
+    final_accuracy = None
```
Comment (kprokofi) [inline torchreid/apis/training.py:111]: in the case when the NNCF is enabled this function will return None for final accuracy. Is it OK?

File: torchreid/integration/sc/train_task.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        :return: None
+        """
+        logger.info("Cancel training requested.")
+        self.stop_callback.stop()
+
+    def save_model(self, output_model: ModelEntity):
+        aux_model_ckpt_dict = {}
+        for name, path in self._aux_model_snap_paths:
+            aux_model_ckpt_dict[name] = load_checkpoint(path, map_location='cpu')
+        state_dict = {'aux_models': aux_model_ckpt_dict}
```
Comment (kprokofi) [inline torchreid/integration/sc/train_task.py]: As I can see from this implementation, state_dict will always be not None. Empty in case of absence of aux models, but not None. In save_model function we compare it with None -> this will lead to updating empty dict

File: torchreid/integration/sc/train_task.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        Called when the user wants to abort training.
+        In this example, this is not implemented.
+
+        :return: None
+        """
+        logger.info("Cancel training requested.")
+        self.stop_callback.stop()
+
+    def save_model(self, output_model: ModelEntity):
```
Comment (kprokofi) [inline torchreid/integration/sc/train_task.py:58]: Is it possible to rename this save_model to _save_model, just to encapsulate it and not be misled with external function?

File: configs/ote_custom_classification/efficientnet_b0/compression_config.json (hunk 2 above)
Comment (alexsu52) [reply configs/ote_custom_classification/efficientnet_b0/compression_config.json]: Thanks! I will delete it.

File: torchreid/__init__.py (hunk 4 above)
Comment (alexsu52) [reply torchreid/__init__.py:3]: I deleted 'import torch'. But I must import NNCF here because NNCF must be imported before calling any torch function.

File: torchreid/apis/training.py (hunk 5 above)
Comment (alexsu52) [reply torchreid/apis/training.py:111]: I think yes. The final accuracy of the NNCF model is not used anywhere. The accuracy of the optimized IR is reported.

File: torchreid/integration/sc/train_task.py (hunk 7 above)
Comment (alexsu52) [reply torchreid/integration/sc/train_task.py:58]: The save_model function is part of the ITrainingTask interface and should be implemented in the OTEClassificationTrainingTask. I can create the common _save_model function in the OTEClassificationInferenceTask. What do you think?

Comment (kprokofi) [reply torchreid/integration/sc/train_task.py:58]: It is not critical. Just not very pretty that function and method have the same name. We may leave it as is or could change the function name instead (which defined in `sc/utils`). `save_model` -> `save_output_model`. It will not interfere with the interface.

**Comment by kprokofi** [comment]
@alexsu52 , please, check the multilabel setup afterward. Let's make sure that it is runnable at least

File: torchreid/integration/nncf/compression_script_utils.py [hunk 8]
```diff
@@ -1,165 +1,153 @@
...
-    nncf_training_config = cfg.get('nncf', {})
-    nncf_config_path = nncf_training_config.get('nncf_config_path')
-    print(f'NNCF config path = {nncf_config_path}')
-    checkpoint_path = cfg.model.load_weights
-    datamanager_for_nncf = None
     _, model, _ = \
-            wrap_nncf_model(model, cfg, datamanager_for_nncf,
-                            checkpoint_path=checkpoint_path,
-                            nncf_config_path=nncf_config_path)
+            wrap_nncf_model(model, cfg)
```
Comment (AlexanderDokuchaev) [inline torchreid/integration/nncf/compression_script_utils.py]: now it can be write in one line

File: torchreid/integration/sc/train_task.py (hunk 6 above)
Comment (alexsu52) [reply torchreid/integration/sc/train_task.py]: Done

File: torchreid/integration/sc/train_task.py (hunk 7 above)
Comment (alexsu52) [reply torchreid/integration/sc/train_task.py:58]: Please, take a look.

File: tools/ote_sample.py (hunk 3 above)
Comment (alexsu52) [reply tools/ote_sample.py]: Please, take a look.

File: torchreid/integration/nncf/compression_script_utils.py [hunk 9]
```diff
@@ -1,165 +1,153 @@
...
+from torchreid.utils import load_checkpoint, load_pretrained_weights
+
+logger = logging.getLogger(__name__)
 
-from torchreid.integration.nncf.compression import (is_checkpoint_nncf, wrap_nncf_model)
 
 def get_coeff_decrease_lr_for_nncf(nncf_training_config):
     if nncf_training_config and nncf_training_config.get('coeff_decrease_lr_for_nncf'):
-        return nncf_training_config.get('coeff_decrease_lr_for_nncf')
+        coef = nncf_training_config.get('coeff_decrease_lr_for_nncf')
```
Comment (AlexanderDokuchaev) [inline torchreid/integration/nncf/compression_script_utils.py]: Double check for `coeff_decrease_lr_for_nncf`

File: torchreid/integration/nncf/engine.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    def dec_train(func):
+        def inner(*args, **kwargs):
+            engine.epoch = kwargs['epoch']
+            func()
+            return
+
+        return inner
+
+    engine.train_data_loader = engine.train_loader
+    engine.max_epoch = nncf_config.get('accuracy_aware_training').get('params').get('maximal_total_epochs', 200)
```
Comment (AlexanderDokuchaev) [inline torchreid/integration/nncf/engine.py]: Not sure that default value needs for `maximal_total_epochs`.
For other parameters like `lr_finder:disable`, `accuracy_drop` should be set manually in config

File: torchreid/integration/sc/configuration.yaml [hunk 11]
```diff
@@ -99,3 +99,55 @@ pot_parameters:
...
+    header: Maximum accuracy degradation
+    max_value: 100.0
+    min_value: 0.0
+    type: FLOAT
+    ui_rules:
+      action: DISABLE_EDITING
+      operator: AND
+      rules: []
+      type: UI_RULES
+    value: 1.0
```
Comment (AlexanderDokuchaev) [inline torchreid/integration/sc/configuration.yaml:149]: `default_value`, `value` should be `0.1`.

File: torchreid/integration/sc/parameters.py [hunk 12]
```diff
@@ -56,6 +70,34 @@ class __LearningParameters(ParameterGroup):
...
+
+        enable_pruning = configurable_boolean(
+            default_value=False,
+            header="Enable filter pruning algorithm",
+            description="Enable filter pruning algorithm",
+            affects_outcome_of=ModelLifecycle.TRAINING
+        )
+
+        maximal_accuracy_degradation = configurable_float(
+            default_value=1.0,
```
Comment (AlexanderDokuchaev) [inline torchreid/integration/sc/parameters.py:93]: 0.1

File: torchreid/integration/sc/configuration.yaml (hunk 11 above)
Comment (alexsu52) [reply torchreid/integration/sc/configuration.yaml:149]: Thanks for the comment! I changed the default value in the templates.

File: torchreid/integration/sc/parameters.py (hunk 12 above)
Comment (alexsu52) [reply torchreid/integration/sc/parameters.py:93]: Thanks for the comment! I changed the default value in the templates.

**Comment by alexsu52** [comment]
> @alexsu52 , please, check the multilabel setup afterward. Let's make sure that it is runnable at least

@kprokofi I pushed changes for the multilabel scenario.

File: torchreid/engine/builder.py [hunk 13]
```diff
@@ -8,7 +8,7 @@ def build_engine(cfg, datamanager, model, optimizer, scheduler,
                  compression_ctrl=None,
                  initial_lr=None):
     if should_freeze_aux_models or nncf_metainfo:
-        if cfg.loss.name not in ['softmax', 'am_softmax']:
+        if cfg.loss.name not in ['softmax', 'am_softmax', 'am_binary']:
             raise NotImplementedError('Freezing of aux models or NNCF compression are supported only for '
                                       'softmax and am_softmax losses for data.type = image')
```
Comment (AlexanderDokuchaev) [inline torchreid/engine/builder.py]:
```suggestion
                                      'softmax, am_softmax and am_binary losses for data.type = image')
```

**Review by AlexanderDokuchaev — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]

**Comment by kprokofi** [comment]
@alexsu52 , is it ready for merge?

**Comment by alexsu52** [comment]
> @alexsu52 , is it ready for merge?

@kprokofi, yes, please, merge
